pkgs/build-support/dlang/builddubpackage/default.nix

```diff
@@ -0,0 +1,129 @@
+{
+  lib,
+  stdenv,
+  fetchurl,
+  linkFarm,
+  dub,
+  ldc,
+  removeReferencesTo,
+}:
+
+# TODO: create proper documentation
+# Until then if you have any questions you can ping the maintainers
+# listed inside `pkgs/build-support/dlang/README.md`
+
+{
+  # path to the lockfile generated by `dub-to-nix`
+  # it can also take in the attrset already processed by lib.importJSON
+  dubLock,
+  # the build type to pass to `dub build`
+  dubBuildType ? "release",
+  # the flags to pass both to `dub build` and `dub test`
+  dubFlags ? [ ],
+  # the flags to pass to `dub build`
+  dubBuildFlags ? [ ],
+  # the flags to pass to `dub test`
+  dubTestFlags ? [ ],
+  # the compiler package used for building
+  compiler ? ldc,
+  ...
+}@args:
+
+let
+  makeDubDep =
+    {
+      pname,
+      version,
+      sha256,
+    }:
+    {
+      inherit pname version;
+      src = fetchurl {
+        name = "dub-${pname}-${version}.zip";
+        url = "mirror://dub/${pname}/${version}.zip";
+        inherit sha256;
+      };
+    };
+
+  lockJson = if lib.isPath dubLock then lib.importJSON dubLock else dubLock;
+
+  lockedDeps = lib.mapAttrsToList (
+    pname: { version, sha256 }: makeDubDep { inherit pname version sha256; }
+  ) lockJson.dependencies;
+
+  # a directory with multiple single element registries
+  # one big directory with all .zip files leads to version parsing errors
+  # when the name of a package is a prefix of the name of another package
+  dubRegistryBase = linkFarm "dub-registry-base" (
+    map (dep: {
+      name = "${dep.pname}/${dep.pname}-${dep.version}.zip";
+      path = dep.src;
+    }) lockedDeps
+  );
+
+  combinedFlags = "--skip-registry=all --compiler=${lib.getExe compiler} ${toString dubFlags}";
+  combinedBuildFlags = "${combinedFlags} --build=${dubBuildType} ${toString dubBuildFlags}";
+  combinedTestFlags = "${combinedFlags} ${toString dubTestFlags}";
+in
+stdenv.mkDerivation (
+  builtins.removeAttrs args [
+    "dubDeps"
+    "extraDubDeps"
```

Comment:
I think you forgot to remove extraDubDeps here

Reply:
Yeah, I forgot about this completely. Now only `dubLock` is in the list.
However, the other inputs will get put into the builder as environment variables (if `__structuredAttrs` is not enabled). Should those also be filtered out?

Or maybe we could keep them and use `finalAttrs.xyz` to reference them, so that we can do `overrideAttrs` on the package. Not sure about how much builders should support overriding ATM. I've read about a builder helper unification plan, but I think it's not going to happen soon.